<a href="https://colab.research.google.com/github/TrevorSeestedt/CSCE580-Fall2024-TrevorSeestedt-Repo/blob/main/Quiz1_Response_Q4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Question 4: Data Analysis for Social Impact
The following is a report of real student driving data in 2023 and 2024.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/DriverTraining-ForInClassLearning-2024.csv')

missing_percentage = (df.isnull().sum() / len(df)) * 100
print(missing_percentage)

class_date    0.000000
city          0.035503
state         0.035503
zip_code      0.035503
dob           0.786982
gender        1.337278
dtype: float64


a. Data Issues
1. As it is seen, the amount of data missing per column is as follows: class_date has 0% missing, city has .0356% missing, state has .0356% missing, zip code has .0356% missing, date of birth ahs .787% missing, and gender has 1.337% missing.
2. There are a few data issues that have occurred in the given data set. The first is the missing values, as there is already a significant amount of data within the set, removing the rows of the missing will not significantly impact the set. The next issue is the formatting of the data, not all cities have been correctly input (some are all lower/upper cased) which may cause confusion as letters can look similar like 'I'(upper case 'i') and 'l'(lower case 'L') and the formatting of dates for the class and birth are the same. If the formatting is changed for the dates then there will be less probability of error as it will stick out if input wrong.

In [ ]:
# create IDs
df['id'] = range(1, len(df) + 1)

In [ ]:
# Removal of rows with missing data
df.dropna(inplace=True)

In [ ]:
# Date format change for class and dob
df['class_date'] = pd.to_datetime(df['class_date'])

In [ ]:
# Convert 'dob' to datetime, replacing invalid dates with NaT
df['dob'] = pd.to_datetime(df['dob'], errors='coerce')

# Remove rows with NaT values in 'dob'
df.dropna(subset=['dob'], inplace=True)

# Now you can apply strftime
df['dob'] = df['dob'].dt.strftime('%B %d, %Y')

<ipython-input-26-d05ea3a6748f>:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['dob'] = pd.to_datetime(df['dob'], errors='coerce')


In [ ]:
# Capitalize cities
df['city'] = df['city'].str.title()

In [ ]:
df.head(15)

class_date            city           state    zip_code                 dob  \
0  2024-08-17        Richburg  South Carolina       29729   November 30, 2007   
1  2024-08-10        Bluffton  South Carolina  29910-7957    October 16, 2007   
2  2024-08-14         Hampton  South Carolina       29924       July 03, 2008   
3  2024-08-09        Columbia  South Carolina       29210    October 10, 2006   
4  2024-08-17            Irmo  South Carolina       29063   December 23, 2006   
5  2024-08-10       Rock Hill  South Carolina       29730  September 03, 2008   
6  2024-08-10  Mount Pleasant  South Carolina       29466   February 14, 2008   
7  2024-08-15      Greenville  South Carolina       29607    January 18, 2007   
8  2024-08-10         Clemson  South Carolina       29631  September 30, 2008   
9  2024-08-15      Blyethwood  South Carolina       29016   December 04, 2005   
10 2024-08-14          Easley  South Carolina       29642       June 03, 2008   
11 2024-08-15         Awendaw  South Carolina       29429     August 08, 2008   
12 2024-08-15  Travelers Rest  South Carolina       29690   December 04, 2007   
13 2024-08-15       Rock Hill  South Carolina       29732       June 03, 2007   
14 2024-08-15            Irmo  South Carolina       29063    January 16, 2007   

    gender  id  
0     Male   1  
1     Male   2  
2   Female   3  
3   Female   4  
4   Female   5  
5   Female   6  
6   Female   7  
7   Female   8  
8   Female   9  
9     Male  10  
10  Female  11  
11  Female  12  
12  Female  13  
13    Male  14  
14  Female  15

In [ ]:
# % of males and females taking the class
gender_percent = df['gender'].value_counts(normalize=True) * 100
print(gender_percent)

gender
Male      50.839676
Female    49.160324
Name: proportion, dtype: float64


In [ ]:
# get the month from the class date
df['month'] = df['class_date'].dt.month

In [ ]:
# gender percent of each month
gender_percent_month = df.groupby('month')['gender'].value_counts(normalize=True) * 100
print(gender_percent_month)

month  gender
1      Female    50.390625
       Male      49.609375
2      Male      52.100840
       Female    47.899160
3      Male      51.108871
       Female    48.891129
4      Male      51.872872
       Female    48.127128
5      Male      50.000000
       Female    50.000000
6      Female    52.251023
       Male      47.748977
7      Female    50.979880
       Male      49.020120
8      Male      52.217916
       Female    47.782084
9      Male      51.355014
       Female    48.644986
10     Male      57.200000
       Female    42.800000
11     Female    51.048951
       Male      48.951049
12     Male      54.265403
       Female    45.734597
Name: proportion, dtype: float64


b. Exploratory Data Analysis - Demographics:
1. As shown, the percent of males is slightly higher than the percent of females taking the class at 50.86% vs 49.14%. Yes, the percentage does vary month by month as the male percentage in october is 57% vs 47% in June.

In [ ]:
from datetime import date

def calc_age (born, today_date):
    return today_date.year - born.year - ((today_date.month, today_date.day) < (born.month, born.day)) # today date - born date

today = date(2024, 5, 8)
df['age'] = df['dob'].apply(lambda x: calc_age(pd.to_datetime(x), today))

In [ ]:
print(f"Overall average age: {df['age'].mean()}")
average_age_by_gender = df.groupby('gender')['age'].mean()
print(average_age_by_gender)

Overall average age: 16.151745801618944
gender
Female    16.137626
Male      16.165399
Name: age, dtype: float64


The average age of the students taking the driving class is 16.1517, with males being slightly older than females at 16.165 (males) and 16.1376 (female). Some assumptions made for the analysis would be that we assume that todays date is September 16th, 2024. Another assumption made would be that the class date is the starting date of the student taking the driving class.

In [ ]:
# percentage of students taking class by cities
city_percent = df['city'].value_counts(normalize=True) * 100
print(city_percent)

city
Lexington         10.064033
Columbia           9.266643
Fort Mill          8.547783
Mount Pleasant     5.768998
Rock Hill          4.784342
                    ...    
Blythewood,Sc      0.006041
Rock Hill Sc       0.006041
West Colimbia      0.006041
Milton             0.006041
Villa Rica         0.006041
Name: proportion, Length: 412, dtype: float64
